### ISSUE TITLE
About the NotImplementedError exception for Lint/UnusedMethodArgument

### ISSUE THREAD

**Issue by sandstrom** [rubocop/rubocop#11393]
First of all, thanks for an awesome project! 🏅 

The current implementation will not complain about unused method arguments if the method raises a `NotImplementedError` (it was implemented in this PR: https://github.com/rubocop/rubocop/issues/7739).

However, many recommend against raising this exception in your code:

- https://oleg0potapov.medium.com/ruby-notimplementederror-dont-use-it-dff1fd7228e5
- https://blog.airbrake.io/blog/ruby-exception-handling/notimplementederror
- http://chrisstump.online/2016/03/23/stop-abusing-notimplementederror/

These three posts all boil down to:

1. Its actual meaning is different than the name implies – [according to RubyDoc](https://ruby-doc.org/core-3.0.0/NotImplementedError.html) it should be raised when a feature is not implemented on the current platform, think of low-level things like calling fork or fsync.
2. It's a descendant of `ScriptError` so a standard rescue block will not capture this, which may cause unexpected problems.

There is also ongoing discussion in this sibling repo to Rubocop:

- https://github.com/rubocop/ruby-style-guide/issues/458

### Downside

It's somewhat weird to have `def read_db(_name)`, where the real method name is `name`.

### Possible solution

- Add an option called `IgnoredExceptions` (array type), with the default value of `['NotImplementedError']`.
  - For anyone who hasn't configured this cop, it's 100% backwards compatible.
    - People who have set `IgnoreNotImplementedMethods` to `false` would need to update their config to use `IgnoredExceptions: []` intead.
  - Could then deprecate the current option called `IgnoreNotImplementedMethods` and remove it in the next major.
- Allow the yarddoc `@abstract` keyword to be used as another signal to disable this lint check on a method.
- Remove the exception case for `NotImplementedError` (if you think the reasons for doing so makes sense).
- Do nothing and consider this an edge case.

### Docs for Lint/UnusedMethodArgument

![image](https://user-images.githubusercontent.com/122287/210799352-992d97da-8614-458d-86cd-259ae5384915.png)

### Addendum

Also, I realize this is nuanced. So feel free to close this issue if you think it's irrelevant or that the current logic is "good enough".

**Comment by dvandersluis** [comment]
Given your sources I think it would make sense to extend the exception to other error classes as well. 

I think it would make sense to augment the existing configuration with a list of allowable exception classes, so something like (abridged):

```yaml
Lint/UnusedMethodArgument:
  IgnoreNotImplementedMethods: true
  NotImplementedExceptions:
    - NotImplementedError
    - NoMethodError # it probably makes sense to include this by default
```

That way no one's configuration would need to change (+/- adding another exception class and if people don't want it). It'd also be easy to extend via configuration if someone was to define their own error to be used for abstract methods.

**Comment by dvandersluis** [comment]
Thinking and reading on this more I think actually we shouldn't add `NoMethodError` by default but if codebases want to use that for marking something not implemented it could be added to the configuration.

**Comment by sandstrom** [comment]
> Thinking and reading on this more I think actually we shouldn't add `NoMethodError` by default but if codebases want to use that for marking something not implemented it could be added to the configuration.

Yeah, I think `NoMethodError` is for something else, I wouldn't include it.

**Comment by sandstrom** [comment]
Awesome! 🎉
